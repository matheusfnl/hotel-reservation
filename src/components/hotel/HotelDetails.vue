<template>
  <q-card-section class="q-pb-none">
    <div class="q-ma-none q-mb-sm row items-center q-gutter-x-sm">
      {{ rating }}

      <q-rating v-model="rating" readonly size="18px" :max="5" color="yellow" />
    </div>

    <p class="text-h4 q-mb-none two-lines-truncate">{{ hotel.getName }}</p>
    <p class="text-grey-6 q-mb-none">{{ hotel.getFullHotelLocation }}</p>
  </q-card-section>

  <q-card-section>
    <q-carousel
      v-model="slide"
      animated
      infinite
      :autoplay="100000"
      arrows
      height="512px"
      class="no-scroll"
      navigation-icon="radio_button_unchecked"
      control-type="regular"
      control-color="white"
      control-text-color="grey"
    >
      <q-carousel-slide
        v-for="(image, index) in hotel.getImages"
        :key="index"
        :name="index"
        class="q-pa-none"
        style="height: 512px"
      >
        <q-img
          :src="image"
          spinner-color="primary"
          fit="cover"
          height="512px"
          class="full-width"
          no-native-menu
        />
      </q-carousel-slide>

      <template v-slot:control>
        <div class="image-counter text-white q-px-md q-py-xs">
          {{ slide + 1 }}/{{ hotel.getImages.length }}
        </div>
      </template>
    </q-carousel>
  </q-card-section>

  <q-card-section v-if="hotel.getAmenities.length">
    <p class="text-h5">{{ $t('hotel.drawer.facilities.title') }}</p>

    <q-separator />

    <div class="q-my-lg amenities-grid">
      <span v-for="amenity in getAmenities" class="amenity-item text-grey-6" :key="amenity.key">
        <q-icon :name="Amenities[amenity.key]" size="sm" />
        {{ amenity.label }}
      </span>
    </div>

    <div class="row justify-center q-mb-lg" v-if="hotel.getAmenities.length > 4">
      <q-btn rounded no-caps color="primary" outline @click="toggleFacilities">
        {{ getAmenitiesText }}
      </q-btn>
    </div>

    <q-separator />
  </q-card-section>

  <q-card-section>
    <p class="text-h5">{{ $t('hotel.drawer.description.title') }}</p>

    <p class="text-grey-6">
      {{ hotel.getDescription }}
    </p>
  </q-card-section>
</template>

<script lang="ts" setup>
import { onMounted, ref, computed } from 'vue'
import { useI18n } from 'vue-i18n'

import type { Hotel } from '@/models/Hotel'

import { Amenities } from '@/enums/amenities'

const { hotel } = defineProps<{
  hotel: Hotel
}>()

const { t } = useI18n()

const rating = ref(0)
const slide = ref(0)
const showAllAmenities = ref(false)

const getAmenities = computed(() => {
  const quantity = showAllAmenities.value ? hotel.getAmenities.length : 4
  const amenities = hotel.getAmenities

  return amenities.splice(0, quantity)
})

const getAmenitiesText = computed(() => {
  return showAllAmenities.value ? t('hotel.drawer.button.less') : t('hotel.drawer.button.more')
})

const toggleFacilities = () => {
  showAllAmenities.value = !showAllAmenities.value
}

onMounted(() => {
  rating.value = +hotel.getRating
})
</script>

<style lang="scss" scoped>
:deep(.q-carousel) {
  border-radius: 16px;
  overflow: hidden;

  .q-carousel__slide {
    border-radius: 16px;
  }

  .q-img {
    border-radius: 16px;
  }

  .q-carousel__prev-arrow--horizontal {
    left: 8px;
  }

  .q-carousel__next-arrow--horizontal {
    right: 8px;
  }

  .q-carousel__arrow {
    button {
      width: 18px;
      height: 18px;
      min-width: unset;
      min-height: unset;

      i {
        font-size: 1rem;
      }
    }
  }
}

.image-counter {
  position: absolute;
  bottom: 16px;
  left: 50%;
  background: rgba(0, 0, 0, 0.7);
  border-radius: 16px;
  transform: translateX(-50%);
}

.amenities-grid {
  display: grid;
  grid-template-columns: repeat(4, auto);
  justify-content: space-between;
}

.amenity-item {
  display: flex;
  align-items: center;
  gap: 8px;
  font-size: 16px;
}

.two-lines-truncate {
  display: -webkit-box;
  line-clamp: 2;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
}
</style>
