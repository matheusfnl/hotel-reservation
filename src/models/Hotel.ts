import { createI18n } from 'vue-i18n'
import messages from '@/i18n'

import type { Amenities, Hotel as HotelInterface } from '@/types'
import { formatValue } from '@/helpers/number-to-value'

const i18n = createI18n({
  locale: 'pt-BR',
  legacy: false,
  messages,
})

export class Hotel {
  private id: number
  private favorite: boolean
  private name: string
  private description: string
  private stars: string
  private thumb: string
  private amenities: Amenities[] | null
  private hasBreakFast: boolean
  private hasRefundableRoom: boolean
  private hasAgreement: boolean
  private nonRefundable: boolean | null
  private address: {
    fullAddress: string
    city: string
    district: string
    street: string
    state: string
    zipCode: string | null
  }
  private images: string[]
  private deals: unknown
  private roomsQuantity: number
  private price: number

  constructor(data: HotelInterface) {
    this.id = data.id
    this.favorite = data.favorite
    this.name = data.name
    this.description = data.description
    this.stars = data.stars
    this.thumb = data.thumb
    this.amenities = data.amenities
    this.hasBreakFast = data.hasBreakFast
    this.hasRefundableRoom = data.hasRefundableRoom
    this.hasAgreement = data.hasAgreement
    this.nonRefundable = data.nonRefundable
    this.address = data.address
    this.images = data.images
    this.deals = data.deals
    this.roomsQuantity = data.roomsQuantity
    this.price = data.price
  }

  public get getId(): number {
    return this.id
  }

  public get getName(): string {
    return this.name
  }

  public get getDescription(): string {
    return this.description
  }

  public get getImages(): string[] {
    return this.images
  }

  public get getHotelLocation(): string {
    const { city, district, street } = this.address

    return `${city}, ${district}, ${street}.`
  }

  public get getFullHotelLocation(): string {
    const { fullAddress, city, state, zipCode } = this.address

    return `${fullAddress}, ${city} - ${state}${zipCode ? `, ${zipCode}` : ''}`
  }

  public get getRating(): string {
    return this.stars
  }

  public get getAmenities(): Amenities[] {
    let amenities = this.amenities ? [...this.amenities] : []
    amenities = amenities.filter((amenity) => !Array.isArray(amenity))

    if (this.hasBreakFast && !amenities.find((amenity) => amenity.key === 'BREAKFAST')) {
      amenities.push({
        key: 'BREAKFAST',
        label: 'Café da manhã',
      })
    }

    return amenities || []
  }

  public get isRefundable(): boolean {
    return this.hasRefundableRoom && !this.nonRefundable
  }

  public get getFullPrice(): string {
    return formatValue(this.price)
  }

  public get getPerNightText(): string {
    const value = this.price / 3

    return `${i18n.global.t('hotel.hotel.currency')}${formatValue(value)}${i18n.global.t('hotel.hotel.per_night')}`
  }
}
