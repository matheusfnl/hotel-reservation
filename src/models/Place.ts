import type { Places, PlaceState } from '@/types'

export class Place {
  readonly name: string
  readonly state: PlaceState
  readonly placeId: number

  constructor(data: Places) {
    this.name = data.name
    this.state = data.state
    this.placeId = data.placeId
  }
}
