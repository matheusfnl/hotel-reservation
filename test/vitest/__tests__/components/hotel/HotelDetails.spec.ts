import { describe, it, expect, vi } from 'vitest'
import { nextTick } from 'vue'
import { QRating } from 'quasar'
import { mount } from '@vue/test-utils'

import HotelDetails from '@/components/hotel/HotelDetails.vue'

import type { Hotel as HotelInterface } from '@/types'
import { Hotel } from '@/models/Hotel'

const mockHotelData: HotelInterface = {
  id: 1,
  favorite: false,
  name: 'Hotel Teste',
  description: 'Descrição do hotel',
  stars: '4.5',
  thumb: 'thumb.jpg',
  amenities: [{ key: 'ROOM_SERVICE', label: 'Wi-Fi' }],
  hasBreakFast: true,
  hasRefundableRoom: true,
  hasAgreement: false,
  nonRefundable: null,
  address: {
    fullAddress: 'Rua Exemplo, 123',
    city: 'São Paulo',
    district: 'Centro',
    street: 'Rua Exemplo',
    state: 'SP',
    zipCode: '12345-678',
  },
  images: ['image1.jpg', 'image2.jpg'],
  deals: null,
  roomsQuantity: 10,
  price: 1000,
}

const mockHotel = new Hotel(mockHotelData)

const mockDrawer = {
  open: vi.fn(),
}

const mountWrapper = () => {
  return mount(HotelDetails, {
    props: {
      hotel: mockHotel,
    },
    global: {
      provide: {
        drawer: mockDrawer,
      },
    },
  })
}

describe('HotelCard', () => {
  it('should render correctly', () => {
    const wrapper = mountWrapper()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should show rating correctly', async () => {
    const wrapper = mountWrapper()
    const rating = wrapper.findComponent(QRating)

    await nextTick()

    expect(rating.props().modelValue).toBe(4.5)
  })
})
