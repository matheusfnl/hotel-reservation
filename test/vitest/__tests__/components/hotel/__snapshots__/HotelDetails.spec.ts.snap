// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelCard > should render correctly 1`] = `
"<div data-v-c52ce958="" class="q-card__section q-card__section--vert q-pb-none">
  <div data-v-c52ce958="" class="row justify-between full-width items-center">
    <div data-v-c52ce958="" class="q-ma-none q-mb-sm row items-center q-gutter-x-sm">0 <div data-v-c52ce958="" class="q-rating row inline items-center q-rating--non-editable text-yellow" style="font-size: 18px;" role="radiogroup" aria-readonly="true">
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 1"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 2"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 3"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 4"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
        <div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 5"><i class="q-icon notranslate material-icons q-rating__icon text-yellow" aria-hidden="true" role="presentation">grade</i></div>
      </div>
    </div>
    <!--v-if-->
  </div>
  <p data-v-c52ce958="" class="text-h4 q-mb-none two-lines-truncate">Hotel Teste</p>
  <p data-v-c52ce958="" class="text-grey-6 q-mb-none">Rua Exemplo, 123, São Paulo - SP, 12345-678</p>
</div>
<div data-v-c52ce958="" class="q-card__section q-card__section--vert">
  <div data-v-c52ce958="" class="q-carousel q-panel-parent q-carousel--without-padding q-carousel--arrows-horizontal no-scroll" style="height: 512px;">
    <div class="q-carousel__slides-container">
      <transition-stub appear="false" persisted="false" css="true">
        <div class="q-panel scroll" style="--q-transition-duration: 300ms;" role="tabpanel">
          <div data-v-c52ce958="" class="q-carousel__slide q-pa-none" style="height: 512px;">
            <div data-v-c52ce958="" class="q-img q-img--no-menu full-width" style="height: 512px;" role="img">
              <div style="padding-bottom: 56.24929688378895%;"></div>
              <div class="q-img__container absolute-full"><img class="q-img__image q-img__image--with-transition q-img__image--current" style="object-fit: cover; object-position: 50% 50%;" height="512px" loading="lazy" fetchpriority="auto" aria-hidden="true" draggable="false" src="image1.jpg"></div>
              <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
                <div class="q-img__loading absolute-full flex flex-center"><svg class="q-spinner text-primary q-spinner-mat" width="1em" height="1em" viewBox="25 25 50 50">
                    <circle class="path" cx="50" cy="50" r="20" fill="none" stroke="currentColor" stroke-width="5" stroke-miterlimit="10"></circle>
                  </svg></div>
              </transition-stub>
            </div>
          </div>
        </div>
      </transition-stub>
    </div>
    <div class="q-carousel__control q-carousel__arrow q-carousel__prev-arrow q-carousel__prev-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-grey q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_left</i></span></button></div>
    <div class="q-carousel__control q-carousel__arrow q-carousel__next-arrow q-carousel__next-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-grey q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_right</i></span></button></div>
    <div data-v-c52ce958="" class="image-counter text-white q-px-md q-py-xs">1/2</div>
  </div>
</div>
<div data-v-c52ce958="" class="q-card__section q-card__section--vert">
  <p data-v-c52ce958="" class="text-h5">Facilidades do hotel</p>
  <hr data-v-c52ce958="" class="q-separator q-separator--horizontal" aria-orientation="horizontal">
  <div data-v-c52ce958="" class="q-my-lg amenities-grid"><span data-v-c52ce958="" class="amenity-item text-grey-6"><i data-v-c52ce958="" class="q-icon notranslate material-icons" style="font-size: 24px;" aria-hidden="true" role="presentation">room_service</i> Wi-Fi</span><span data-v-c52ce958="" class="amenity-item text-grey-6"><i data-v-c52ce958="" class="q-icon notranslate material-icons" style="font-size: 24px;" aria-hidden="true" role="presentation">bakery_dining</i> Café da manhã</span></div>
  <!--v-if-->
  <hr data-v-c52ce958="" class="q-separator q-separator--horizontal" aria-orientation="horizontal">
</div>
<div data-v-c52ce958="" class="q-card__section q-card__section--vert">
  <p data-v-c52ce958="" class="text-h5">Conheça um pouco mais</p>
  <p data-v-c52ce958="" class="text-grey-6">Descrição do hotel</p>
</div>"
`;
